File: UVa/MonkeyOiledBamboo12032.py
def simulate(k):
    barely = False
    for i in range(len(rugs) - 1):
        if rugs[i + 1] - rugs[i] > k:
            return False, False
        elif rugs[i + 1] - rugs[i] == k:
            k -= 1
            barely = True
    return barely, True

t = int(input())
for tcase in range(1, t + 1):
    n = int(input())
    rugs = list(map(int, input().split()))
    rugs.insert(0, 0)
    diff = 0
    for i in range(1, len(rugs)):
        print(rugs[i] - rugs[i - 1], end = " ")
        diff = max(diff, rugs[i] - rugs[i - 1])
    print()
    
    lo, hi = diff, pow(10, 7) + 1
    k = (lo + hi) // 2
    barely, success = simulate(k)
    tries = 0
    while not (success and barely):
        print(lo, hi, k, success, barely)

        if tries > 200:
            r = 1/ 0
            
        if not success:
            lo = k + 1
        if success and not barely:
            hi = k

        k = (lo + hi) // 2
        barely, success = simulate(k)
        tries += 1

    print("Case {}: {}".format(tcase, k))
        

File: UVa/MonkeyOiledBamboo12032.cpp
using namespace std;
#define ll long long
#include <bits/stdc++.h>

bool simulate(int k, vector<int> &rugs){
    for (int i = 0; i < rugs.size() - 1; i++){
        if (rugs[i + 1] - rugs[i] > k){
            return false;
        } else if (rugs[i + 1] - rugs[i] == k){
            k -= 1;
        }
    }

    return true;
}

int main(){
    ios::sync_with_stdio(0), cin.tie(0);
    int t;
    cin >> t;
    for (int tcase = 1; tcase <= t; tcase++){
        int n;
        cin >> n;
        vector<int> rugs(n+1);
        rugs[0] = 0; //Floor
        for (int i = 1; i <= n; i++) cin >> rugs[i];

        int lo = 0, hi = pow(10, 7) + 1;
        int k = (lo + hi) / 2;
        int best = hi;
        int tries = 0;
        while (tries <= 24){
            if (simulate(k, rugs)){
                best = min(best, k);
                hi = k - 1;
            } else {
                lo = k + 1;
            }
            tries++;
            k = (lo + hi) / 2;
        }

        cout << "Case " << tcase << ": " << best << "\n";
    }
}